Raise ArgumentError if an unrecognised callback is skipped

iainbeeston opened the pull request:
At present, if you skip a callback that hasn't been defined, activesupport callbacks silently does nothing. However, it's easy to mistype the name of a callback and mistakenly think that it's being skipped, when it is not.

This problem even exists in the current test suite. CallbacksTest::SkipCallbacksTest#test_skip_person attempts to skip callbacks that were never set up.

This PR amends the current behavior by printing a warning if a callback is skipped without having been defined first.

There is one catch to this, which is that `AbstractController::Callbacks#skip_action_callback` (which attempts to skip all before, around and after actions with the specified name) will always show a warning when used (unless the same callback is configured for before, around and after). If this PR is accepted I'd suggest that `skip_action_callback` is deprecated, or modified to check whether a callback exists before removing it (I'd go for the former option).

iainbeeston committed:
Raise ArgumentError if an unrecognised callback is skipped

At present, if you skip a callback that hasn't been defined,
activesupport callbacks silently does nothing. However, it's easy to
mistype the name of a callback and mistakenly think that it's being
skipped, when it is not.

This problem even exists in the current test suite.
CallbacksTest::SkipCallbacksTest#test_skip_person attempts to skip
callbacks that were never set up.

This PR changes `skip_callback` to raise an `ArgumentError` if the
specified callback cannot be found.

egilburg commented on activesupport/lib/active_support/callbacks.rb:
```diff
@@ -693,14 +693,18 @@ def skip_callback(name, *filter_list, &block)
-            if filter && options.any?
-              new_filter = filter.merge_conditional_options(chain, if_option: options[:if], unless_option: options[:unless])
-              chain.insert(chain.index(filter), new_filter)
+            if !callback
```
Better to do `unless callback`

iainbeeston replied:
Good point

egilburg commented on activesupport/lib/active_support/callbacks.rb:
```diff
@@ -693,14 +693,18 @@ def skip_callback(name, *filter_list, &block)
+              warn "#{type.to_s.capitalize} #{name} callback #{filter.inspect} has not been defined"
             end
 
-            chain.delete(filter)
```
You changed order of deleting callback to be after code which checks `index(callback)`. Is this material?

iainbeeston replied:
Sorry, I don't understand. In both cases the delete is at the end.

The git diff isn't very clear, because I've renamed filter to callback (to avoid shadowing the loop variable for filter)

egilburg commented on activesupport/lib/active_support/callbacks.rb:
```diff
@@ -693,14 +693,18 @@ def skip_callback(name, *filter_list, &block)
+              chain.insert(chain.index(callback), new_callback)
+            end
+
+            chain.delete(callback)
```
Should this be in the `if callback` clause?

I think it'll be clearer if code looks like
```ruby
if callback
  if options.any?
    ... 
  end

  ...
else
  warn(...)
end

iainbeeston replied:
I'd argue not, because the intent of the two if statements are totally different and unrelated:

1. The new one that I've added performs error handling (based on whether a callback could be found)
2. The other one transforms an existing callback so that it's only run in certain situations, but only if specific method parameters were used

rafaelfranca commented:
Why not instead of printing a warning we don't raise an ArgumentError?

iainbeeston commented:
@rafaelfranca That's a great question. I'd prefer to do that, however it's a change in behaviour that would cause both the rails test suite and existing apps to break. `AbstractController::Callbacks#skip_action_callback` would also need to be removed (possibly not a bad thing). I didn't want to make that call, but if you'd prefer that I'd be happy to make the change (and fix up the test suite). I admit that not many rails apps run with warnings enabled.

rafaelfranca commented:
> however it's a change in behaviour that would cause both the rails test suite and existing apps to break.

This will be released in a major release, so this will be not a problem if documented properly.

About `skip_action_callback` I'd deprecate and remove it later in favor of a more explicit skip call.

iainbeeston commented:
> This will be released in a major release, so this will be not a problem if documented properly.

Great, I'll do that.

> About skip_action_callback I'd deprecate and remove it later in favor of a more explicit skip call.

Would you like me to do that in a separate pull request? Presumably that change would need to be released before we make `skip_callback` raise `ArgumentError`s, because `skip_action_callback` calls `skip_callback` for all of `:before`, `:after`, and `:around` callback types and (unless the same callback has been set for all three) it will always raise an `ArgumentError`.

rafaelfranca commented:
> Would you like me to do that in a separate pull request?

:+1:

iainbeeston commented:
I've raised #19053 to deprecate `AbstractController::Callbacks#skip_action_callback` (next I will update this branch to raise an `ArgumentError` rather than a warning)

iainbeeston commented:
I've now updated this to raise an ArgumentError (rather than showing a warning).

I've also removed AbstractController::Callbacks#skip_action_callback, as this change makes that method always raise an ArgumentError. Because of that, this change should probably be delayed until the minor release after #19060 (which deprecates that method). Can we tag this PR as targeting rails 5.1?

rafaelfranca commented:
I think we can accept a option that does nothing instead of raising an error, so we can use in `skip_action_callback` until it is not removed.

rafaelfranca commented:
What do you think?

iainbeeston commented:
@rafaelfranca Good compromise. I've updated the code to preserve the original behavior (ie. not raise an ArgumentError) if the `:raise` option is set to `false`.

Please let me know if you'd like me to make further improvements.

iainbeeston commented:
I've rebased to resolve the merge conflict (should be ok to merge now)

iainbeeston commented:
Looks like there's a build failure in master though

iainbeeston commented:
@rafaelfranca Could you please take another look at this? I've incorporated the changes you suggested, the build is passing and there are no merge conflicts. (Thanks!)

iainbeeston commented:
Is there anything else I can do to get this merged?

iainbeeston commented:
I've rebased this again, so it's good to merge
